### ISSUE TITLE
Self-adversarial negative sampling

### ISSUE THREAD

**Issue by Otamio** [facebookresearch/PyTorch-BigGraph#236]
Hi, Luca. Thanks you and your team for your awesome work! I have been doing graph related embedding work, and PBG has been really a handy tool for me to work with large graphs.

One thing I want to know is whether PBG allows users to create new sampling strategies on their own (just like operators and comparators). Currently PBG only supports uniform and same batch sampling. Using this configuration I was only able to get about ~0.25 MRR on the FB15K-237 dataset. Self-adversarial negative sampling was recently introduced (https://arxiv.org/abs/1902.10197), and it appeared to produce significantly better results (~0.3 MRR on FB15K-237). I definitely want to try the sampling method with larger graphs. 

Do you think it possible to customize the sampling strategy based on the PBG framework? If yes, what parts of the program do you think needs modification?

Many thanks!

**Comment by lw** [comment]
Indeed right now there's no "modular plug-in system" for negative sampling strategies (unlike what we have for operators, comparators, ...). @adamlerer has in the past shown interest in trying out other sampling strategies, maybe he can comment with some thoughts?

**Comment by adamlerer** [comment]
Hi @Otamio , I took a quick look at this paper. It looks like in order to find negative samples for a single node u they compute the score for every pair (u, v) for *every node v in the graph* and weight them by the softmax distribution. Such a strategy is fine for a tiny graph like fb15k but doesn't scale to large graphs. This is why we use the "batched negative sampling" trick described in our paper to get a lot of negative samples cheaply, and specifically avoid any "hard negative sampling".

PBG is really better suited to large web interaction graphs than large knowledge graphs; for such graphs, the feedback I've heard is that our batched negative sampling makes hard negative mining unnecessary.

**Comment by Otamio** [comment]
Thanks @adamlerer . I will take a deeper look into the paper and do a couple of experiments to see if sampling strategies matters.
